addons/web/models/models.py

```diff
@@ -472,35 +563,86 @@ def group_id_name(value):
+                        count_domain = AND([
+                            count_domain,
+                            extra_domain,
+                        ])
+                    count = self.search_count(count_domain)
+
+                if expand or count:
+                    if enable_counters:
+                        values['__count'] = count
+                    field_range.append(values)
+
+            return field_range
 
-        if field.type == 'many2many' and kwargs.get('enable_counters'):
-            # fetch the counts
-            model_domain = AND([
-                kwargs.get('search_domain', []),
-                kwargs.get('category_domain', []),
-                kwargs.get('filter_domain', []),
-                [(field_name, '!=', False)]
-            ])
-            group_domain = kwargs.get('group_domain', {})
-            for id in values_range:
-                values = values_range[id]
+        if field.type == 'many2one':
+            domain_image = {}
+            if enable_counters or not expand:
+                model_domain = AND([
+                    model_domain,
+                    kwargs.get('group_domain', []),
+                ])
+                domain_image = self._search_panel_domain_image(field_name, model_domain, enable_counters)
+
+            if not (expand or group_by or comodel_domain):
+                return list(domain_image.values())
+
+            if not expand:
+                image_element_ids = list(domain_image.keys())
+                comodel_domain = AND([
+                    comodel_domain,
+                    [('id', 'in', image_element_ids)],
+                ])
+            comodel_records = Comodel.search_read(comodel_domain, field_names)
+
+            field_range = []
+            for record in comodel_records:
+                record_id = record['id']
+                values= {
+                    'id': record_id,
+                    'display_name': record['display_name'],
+                }
+
+                if group_by:
+                    group_id, group_name = group_id_name(record[group_by])
+                    values['group_id'] = group_id
+                    values['group_name'] = group_name
 
-                count_domain = AND([model_domain, [(field_name, 'in', id)], ])
-                if group_by and group_domain:
-                    group_id = json.dumps(values['group_id'])
-                    extra_domain = group_domain.get(group_id, [])
-                    count_domain = AND([count_domain, extra_domain, ])
+                if enable_counters:
+                    image_element  = domain_image.get(record_id)
+                    values['__count'] = image_element['__count'] if image_element else 0
 
-                values['__count'] = self.search_count(count_domain)
+                field_range.append(values)
 
-        return list(values_range.values())
+            return field_range
```

Comment:
What happens if the field type is not `many2many` or `many2one`?

Reply:
A check is done at the beginning of the function: the field types supported are selection, many2many and many2one for
the case select="multi"